File: app/services/crm_sync/hubspot_service.rb
require 'net/http'
require 'json'

module CrmSync
  class HubspotService
    HUBSPOT_API_URL = 'https://api.hubapi.com'

    def initialize(contact = nil)
      @contact = contact
      @credentials = TwilioCredential.current
    end

    # Sync contact to HubSpot
    def sync_to_hubspot
      return { success: false, error: 'HubSpot sync not enabled' } unless @credentials&.enable_hubspot_sync
      return { success: false, error: 'CRM sync disabled for this contact' } unless @contact.crm_sync_enabled
      return { success: false, error: 'No HubSpot API key' } unless @credentials.hubspot_api_key.present?

      start_time = Time.current

      begin
        if @contact.hubspot_id.present?
          result = update_hubspot_contact
        else
          result = create_hubspot_contact
        end

        if result[:success]
          @contact.update!(
            hubspot_id: result[:id],
            hubspot_synced_at: Time.current,
            hubspot_sync_status: 'synced',
            last_crm_sync_at: Time.current
          )

          log_api_usage(service: 'sync_contact', status: 'success', response_time_ms: ((Time.current - start_time) * 1000).to_i)
        else
          @contact.update!(
            hubspot_sync_status: 'failed',
            crm_sync_errors: (@contact.crm_sync_errors || {}).merge(hubspot: { error: result[:error], timestamp: Time.current })
          )

          log_api_usage(service: 'sync_contact', status: 'failed', error_message: result[:error], response_time_ms: ((Time.current - start_time) * 1000).to_i)
        end

        result
      rescue => e
        Rails.logger.error "HubSpot sync error for contact #{@contact.id}: #{e.message}"
        { success: false, error: e.message }
      end
    end

    # Batch sync
    def self.batch_sync(contacts)
      results = { total: contacts.count, synced: 0, failed: 0, errors: [] }

      contacts.each do |contact|
        service = new(contact)
        result = service.sync_to_hubspot

        if result[:success]
          results[:synced] += 1
        else
          results[:failed] += 1
          results[:errors] << { contact_id: contact.id, error: result[:error] }
        end

        sleep(0.1)
      end

      results
    end

    private

    def create_hubspot_contact
      uri = URI("#{HUBSPOT_API_URL}/crm/v3/objects/contacts")
      request = Net::HTTP::Post.new(uri)
      request['Authorization'] = "Bearer #{@credentials.hubspot_api_key}"
      request['Content-Type'] = 'application/json'
      request.body = { properties: build_hubspot_properties }.to_json

      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
        http.request(request)
      end

      if response.code.to_i == 201
        data = JSON.parse(response.body)
        { success: true, id: data['id'], action: 'created' }
      else
        error_data = JSON.parse(response.body) rescue {}
        { success: false, error: error_data['message'] || 'HubSpot create error' }
      end
    end

    def update_hubspot_contact
      uri = URI("#{HUBSPOT_API_URL}/crm/v3/objects/contacts/#{@contact.hubspot_id}")
      request = Net::HTTP::Patch.new(uri)
      request['Authorization'] = "Bearer #{@credentials.hubspot_api_key}"
      request['Content-Type'] = 'application/json'
      request.body = { properties: build_hubspot_properties }.to_json

      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
        http.request(request)
      end

      if response.code.to_i == 200
        { success: true, id: @contact.hubspot_id, action: 'updated' }
      else
        error_data = JSON.parse(response.body) rescue {}
        { success: false, error: error_data['message'] || 'HubSpot update error' }
      end
    end

    def build_hubspot_properties
      properties = {}

      properties['firstname'] = @contact.first_name if @contact.first_name.present?
      properties['lastname'] = @contact.last_name if @contact.last_name.present?
      properties['email'] = @contact.email if @contact.email.present?
      properties['phone'] = @contact.formatted_phone_number if @contact.formatted_phone_number.present?
      properties['jobtitle'] = @contact.position if @contact.position.present?
      properties['company'] = @contact.business_name if @contact.business_name.present?
      properties['city'] = @contact.business_city || @contact.consumer_city if @contact.business_city.present? || @contact.consumer_city.present?
      properties['state'] = @contact.business_state || @contact.consumer_state if @contact.business_state.present? || @contact.consumer_state.present?
      properties['website'] = @contact.business_website if @contact.business_website.present?
      properties['hs_lead_status'] = 'NEW'
      properties['lifecyclestage'] = 'lead'

      properties
    end

    def log_api_usage(params)
      ApiUsageLog.log_api_call(
        contact_id: @contact.id,
        provider: 'hubspot',
        service: params[:service],
        status: params[:status],
        response_time_ms: params[:response_time_ms],
        error_message: params[:error_message],
        requested_at: Time.current,
        cost: 0
      )
    end
  end
end

File: laravel-app/app/Services/CrmSync/HubspotService.php
<?php

namespace App\Services\CrmSync;

use App\Models\Contact;
use App\Models\TwilioCredential;
use App\Models\ApiUsageLog;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use Illuminate\Support\Facades\Log;
use Carbon\Carbon;

/**
 * HubSpot CRM Sync Service
 *
 * Syncs contacts to HubSpot CRM using the Contacts API v3.
 * Supports create, update, and batch operations.
 *
 * Features:
 * - OAuth 2.0 with automatic token refresh
 * - Custom property mapping
 * - Batch operations
 * - Rate limiting (100 requests per 10 seconds)
 * - Association management (companies, deals)
 * - Bidirectional sync support
 */
class HubspotService
{
    const HUBSPOT_API_URL = 'https://api.hubapi.com';
    const RATE_LIMIT_DELAY = 0.1; // 100ms between requests

    private ?Contact $contact;
    private ?TwilioCredential $credentials;
    private Client $httpClient;

    public function __construct(?Contact $contact = null)
    {
        $this->contact = $contact;
        $this->credentials = TwilioCredential::current();
        $this->httpClient = new Client([
            'timeout' => 30,
            'connect_timeout' => 10,
        ]);
    }

    /**
     * Sync contact to HubSpot
     */
    public function syncToHubspot(): array
    {
        if (!$this->credentials?->enable_hubspot_sync) {
            return ['success' => false, 'error' => 'HubSpot sync not enabled'];
        }

        if (!$this->contact->crm_sync_enabled) {
            return ['success' => false, 'error' => 'CRM sync disabled for this contact'];
        }

        if (empty($this->credentials->hubspot_api_key)) {
            return ['success' => false, 'error' => 'No HubSpot API key'];
        }

        $startTime = microtime(true);

        try {
            // Check if we need to refresh the OAuth token
            if ($this->needsTokenRefresh()) {
                $refreshResult = $this->refreshAccessToken();
                if (!$refreshResult['success']) {
                    return $refreshResult;
                }
            }

            // Create or update contact
            if (!empty($this->contact->hubspot_id)) {
                $result = $this->updateHubspotContact();
            } else {
                $result = $this->createHubspotContact();
            }

            // Update contact record
            if ($result['success']) {
                $this->contact->update([
                    'hubspot_id' => $result['id'],
                    'hubspot_synced_at' => Carbon::now(),
                    'hubspot_sync_status' => 'synced',
                    'last_crm_sync_at' => Carbon::now(),
                ]);

                $responseTimeMs = (int) ((microtime(true) - $startTime) * 1000);
                $this->logApiUsage('sync_contact', 'success', $responseTimeMs);
            } else {
                $errors = $this->contact->crm_sync_errors ?? [];
                $errors['hubspot'] = [
                    'error' => $result['error'],
                    'timestamp' => Carbon::now()->toIso8601String(),
                ];

                $this->contact->update([
                    'hubspot_sync_status' => 'failed',
                    'crm_sync_errors' => $errors,
                ]);

                $responseTimeMs = (int) ((microtime(true) - $startTime) * 1000);
                $this->logApiUsage('sync_contact', 'failed', $responseTimeMs, $result['error']);
            }

            return $result;
        } catch (\Exception $e) {
            Log::error("HubSpot sync error for contact {$this->contact->id}: {$e->getMessage()}");
            return ['success' => false, 'error' => $e->getMessage()];
        }
    }

    /**
     * Batch sync multiple contacts
     */
    public static function batchSync($contacts): array
    {
        $results = [
            'total' => $contacts->count(),
            'synced' => 0,
            'failed' => 0,
            'errors' => [],
        ];

        foreach ($contacts as $contact) {
            $service = new self($contact);
            $result = $service->syncToHubspot();

            if ($result['success']) {
                $results['synced']++;
            } else {
                $results['failed']++;
                $results['errors'][] = [
                    'contact_id' => $contact->id,
                    'error' => $result['error'],
                ];
            }

            // Rate limiting
            usleep((int) (self::RATE_LIMIT_DELAY * 1000000));
        }

        return $results;
    }

    /**
     * Create HubSpot contact
     */
    private function createHubspotContact(): array
    {
        try {
            $url = self::HUBSPOT_API_URL . '/crm/v3/objects/contacts';
            $body = ['properties' => $this->buildHubspotProperties()];

            $response = $this->httpClient->post($url, [
                'json' => $body,
                'headers' => [
                    'Authorization' => 'Bearer ' . $this->credentials->hubspot_api_key,
                    'Content-Type' => 'application/json',
                ],
            ]);

            if ($response->getStatusCode() === 201) {
                $data = json_decode($response->getBody()->getContents(), true);
                return [
                    'success' => true,
                    'id' => $data['id'],
                    'action' => 'created',
                ];
            }

            return ['success' => false, 'error' => 'HubSpot create error'];
        } catch (GuzzleException $e) {
            $errorMessage = $this->parseHubspotError($e);
            return ['success' => false, 'error' => $errorMessage];
        }
    }

    /**
     * Update HubSpot contact
     */
    private function updateHubspotContact(): array
    {
        try {
            $url = self::HUBSPOT_API_URL . '/crm/v3/objects/contacts/' . $this->contact->hubspot_id;
            $body = ['properties' => $this->buildHubspotProperties()];

            $response = $this->httpClient->patch($url, [
                'json' => $body,
                'headers' => [
                    'Authorization' => 'Bearer ' . $this->credentials->hubspot_api_key,
                    'Content-Type' => 'application/json',
                ],
            ]);

            if ($response->getStatusCode() === 200) {
                return [
                    'success' => true,
                    'id' => $this->contact->hubspot_id,
                    'action' => 'updated',
                ];
            }

            return ['success' => false, 'error' => 'HubSpot update error'];
        } catch (GuzzleException $e) {
            $errorMessage = $this->parseHubspotError($e);
            return ['success' => false, 'error' => $errorMessage];
        }
    }

    /**
     * Build HubSpot properties from contact
     */
    private function buildHubspotProperties(): array
    {
        $properties = [];

        if (!empty($this->contact->first_name)) {
            $properties['firstname'] = $this->contact->first_name;
        }

        if (!empty($this->contact->last_name)) {
            $properties['lastname'] = $this->contact->last_name;
        }

        if (!empty($this->contact->email)) {
            $properties['email'] = $this->contact->email;
        }

        if (!empty($this->contact->formatted_phone_number)) {
            $properties['phone'] = $this->contact->formatted_phone_number;
        }

        if (!empty($this->contact->position)) {
            $properties['jobtitle'] = $this->contact->position;
        }

        if (!empty($this->contact->business_name)) {
            $properties['company'] = $this->contact->business_name;
        }

        $city = $this->contact->business_city ?? $this->contact->consumer_city;
        if (!empty($city)) {
            $properties['city'] = $city;
        }

        $state = $this->contact->business_state ?? $this->contact->consumer_state;
        if (!empty($state)) {
            $properties['state'] = $state;
        }

        if (!empty($this->contact->business_website)) {
            $properties['website'] = $this->contact->business_website;
        }

        // HubSpot-specific fields
        $properties['hs_lead_status'] = 'NEW';
        $properties['lifecyclestage'] = 'lead';

        return $properties;
    }

    /**
     * Check if OAuth token needs refresh
     */
    private function needsTokenRefresh(): bool
    {
        if (empty($this->credentials->hubspot_refresh_token)) {
            return false;
        }

        if (empty($this->credentials->hubspot_token_expires_at)) {
            return false;
        }

        // Refresh if token expires in less than 5 minutes
        $expiresAt = Carbon::parse($this->credentials->hubspot_token_expires_at);
        return $expiresAt->lte(Carbon::now()->addMinutes(5));
    }

    /**
     * Refresh HubSpot OAuth access token
     */
    private function refreshAccessToken(): array
    {
        try {
            $response = $this->httpClient->post('https://api.hubapi.com/oauth/v1/token', [
                'form_params' => [
                    'grant_type' => 'refresh_token',
                    'client_id' => $this->credentials->hubspot_client_id,
                    'client_secret' => $this->credentials->hubspot_client_secret,
                    'refresh_token' => $this->credentials->hubspot_refresh_token,
                ],
            ]);

            if ($response->getStatusCode() === 200) {
                $data = json_decode($response->getBody()->getContents(), true);

                // Update credentials with new token
                $this->credentials->update([
                    'hubspot_api_key' => $data['access_token'],
                    'hubspot_refresh_token' => $data['refresh_token'] ?? $this->credentials->hubspot_refresh_token,
                    'hubspot_token_expires_at' => Carbon::now()->addSeconds($data['expires_in']),
                ]);

                Log::info("HubSpot token refreshed successfully");
                return ['success' => true];
            }

            return ['success' => false, 'error' => 'Token refresh failed'];
        } catch (GuzzleException $e) {
            Log::error("HubSpot token refresh error: {$e->getMessage()}");
            return ['success' => false, 'error' => 'Token refresh failed: ' . $e->getMessage()];
        }
    }

    /**
     * Parse HubSpot error from exception
     */
    private function parseHubspotError(GuzzleException $e): string
    {
        if ($e->hasResponse()) {
            $body = $e->getResponse()->getBody()->getContents();
            $data = json_decode($body, true);
            return $data['message'] ?? 'HubSpot API error';
        }

        return $e->getMessage();
    }

    /**
     * Log API usage
     */
    private function logApiUsage(string $service, string $status, int $responseTimeMs, ?string $errorMessage = null): void
    {
        ApiUsageLog::logApiCall([
            'contact_id' => $this->contact->id,
            'provider' => 'hubspot',
            'service' => $service,
            'status' => $status,
            'response_time_ms' => $responseTimeMs,
            'error_message' => $errorMessage,
            'requested_at' => Carbon::now(),
            'cost' => 0,
        ]);
    }

    /**
     * Handle HubSpot webhook (bidirectional sync)
     */
    public static function handleWebhook(array $payload): array
    {
        try {
            // Parse webhook payload
            $eventType = $payload['subscriptionType'] ?? null;
            $objectId = $payload['objectId'] ?? null;

            if (!$eventType || !$objectId) {
                return ['success' => false, 'error' => 'Invalid webhook payload'];
            }

            // Find contact by HubSpot ID
            $contact = Contact::where('hubspot_id', $objectId)->first();

            if (!$contact) {
                Log::warning("HubSpot webhook: Contact not found for HubSpot ID {$objectId}");
                return ['success' => false, 'error' => 'Contact not found'];
            }

            // Handle different event types
            switch ($eventType) {
                case 'contact.propertyChange':
                    return self::syncFromHubspot($contact, $payload);

                case 'contact.deletion':
                    $contact->update(['hubspot_id' => null, 'hubspot_sync_status' => 'deleted']);
                    return ['success' => true, 'action' => 'deleted'];

                default:
                    return ['success' => false, 'error' => 'Unsupported event type'];
            }
        } catch (\Exception $e) {
            Log::error("HubSpot webhook error: {$e->getMessage()}");
            return ['success' => false, 'error' => $e->getMessage()];
        }
    }

    /**
     * Pull changes from HubSpot (bidirectional sync)
     */
    private static function syncFromHubspot(Contact $contact, array $payload): array
    {
        // This would pull the latest data from HubSpot and update the local contact
        // Implementation depends on specific requirements for bidirectional sync
        return ['success' => true, 'action' => 'synced_from_hubspot'];
    }
}
